file_sys: change RomFS archive to Self NCCH archive

wwylele opened the pull request:
aka "Correct FS error codes Part 2"

Addressed #2515. I did a hardware test and confirmed the use of archive 0x00000003 to access ExeFS sections. 

Also fixed the crash when a game tries to access non-exist RomFS (#2519, probably #1862, #1412). Many homebrew applications depend on the RomFS open result to choose whether load from RomFS or SDMC. Now citra returns a proper error code and they should work fine.

Note that ArchiveNCCH (0x2345678A) also have similar ability to access ExeFS section, but its file path format is different from SelfNCCH one (means they cannot share the same `ArchiveBackend` implementation), and it is much more complicated, so I keep it as-is, alone with the legacy `IVFCArchive`, and leave it to a future PR.

All added error codes are verified against real hardware. However I cannot provide the testing program because it is a big mess... if it is need, I will clean it up and upload it later.

A small diagram of the component changes: (green - added; red - removed)
![rom_fs](https://cloud.githubusercontent.com/assets/4592895/22853485/471c5cf2-f060-11e6-800e-51b94c5dde16.png)

Subv commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,239 @@
+    RomFS = 0,
+    Code = 1, // This is not supported by SelfNCCHArchive but by archive 0x2345678E
+    ExeFS = 2,
+    UpdateRomFS = 5,
```
What does the UpdateRomFS do?

wwylele replied:
Presumably for accessing the RomFS of the update patch. 

On 3dbrew it says:
>The update RomFS can be accessed with the first u32 in the binary file-lowpath being set to 0x5.

(though this part of documentation is a big mess)

 On my self-made CIA-installed program without update patch, this gives the same data as RomFS (0)

yuriks replied:
Document this in the code.

Subv commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,239 @@
+
+    std::vector<u8> buffer;
+
+    if (Loader::ResultStatus::Success == app_loader.ReadIcon(buffer))
```
Should we log an error in the failure case of each of these?

mailwl commented:
Now proper EULA version checking from icon may broke some games (mostly which wants ErrEULA applet before).
But we tune it later.

lioncash commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,239 @@
+            LOG_ERROR(Service_FS, "Reading code section is not supported!");
+            return ERROR_COMMAND_NOT_ALLOWED;
+
+        case SelfNCCHFilePathType::ExeFS: {
```
Given the conditionals contain nested returns, `else` can be omitted in quite a few places and simply be:

```cpp
case SelfNCCHFilePathType::ExeFS: {
    std::string filename(file_path.exefs_filename.begin(), file_path.exefs_filename.end());
    const size_t pos = filename.find('\0');
    if (pos != std::string::npos)
        filename.erase(pos);

    if (filename == "icon") {
        if (ncch_data.icon) {
            return MakeResult<std::unique_ptr<FileBackend>>(
                std::make_unique<ExeFSSectionFile>(*ncch_data.icon));
        }

        LOG_ERROR(Service_FS, "Unable to read icon!");
        return ERROR_EXEFS_SECTION_NOT_FOUND;
    }

    if (filename == "logo") {
        if (ncch_data.logo) {
            return MakeResult<std::unique_ptr<FileBackend>>(
                std::make_unique<ExeFSSectionFile>(*ncch_data.logo));
        }

        LOG_ERROR(Service_FS, "Unable to read logo!");
        return ERROR_EXEFS_SECTION_NOT_FOUND;
    }

    if (filename == "banner") {
        if (ncch_data.banner) {
            return MakeResult<std::unique_ptr<FileBackend>>(
                std::make_unique<ExeFSSectionFile>(*ncch_data.banner));
        }

        LOG_ERROR(Service_FS, "Unable to read banner!");
        return ERROR_EXEFS_SECTION_NOT_FOUND;
    }

    LOG_ERROR(Service_FS, "Unknown ExeFS section %s!", filename.c_str());
    return ERROR_INVALID_PATH;
}
```

Another suggestion to consider might be also moving it to a private `OpenExeFS()` function, like with the RomFS code suggested above, so no fallthrough can be introduced by accident if the code ever gets changed.

wwylele replied:
Done

wwylele commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,239 @@
+
+        switch (file_path.type) {
+        case SelfNCCHFilePathType::UpdateRomFS:
+            LOG_WARNING(Service_FS, "(STUBBED) open update RomFS");
```
Yes, it's intentional. Will separate

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+    std::shared_ptr<std::vector<u8>> data;
+};
+
+class SelfNCCHArchive final : public ArchiveBackend {
```
Document what the SelfNCCHArchive does in general and what kinds of "files" it lets you open.

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+
+        case SelfNCCHFilePathType::Code:
+            LOG_ERROR(Service_FS, "Reading code section is not supported!");
+            return ERROR_COMMAND_NOT_ALLOWED;
```
This feels like a misnomer, but I don't have much additional context into the structure of the service. Seems like what's currently called `ERROR_UNSUPPORTED_OPEN_FLAGS` should be comething like `ERROR_UNSUPPORTED_OPERATION`, and then this should be `ERROR_UNSUPPORTED_PATH_TYPE`, but maybe it makes more sense when you take into account other usages.

wwylele replied:
The name `ERROR_COMMAND_NOT_ALLOWED` comes from 3dbrew documentation:
> 630 	Command not allowed / missing permissions? (fs) 

and yeah, I agree that currently not all error codes are named correctly. I prefer using the existing name for a while, and after collecting enough usage of these codes, we then do a name clean up

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+            return OpenRomFS();
+
+        case SelfNCCHFilePathType::Code:
+            LOG_ERROR(Service_FS, "Reading code section is not supported!");
```
Reading *the* code section [...]

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+    ResultVal<std::unique_ptr<FileBackend>> OpenRomFS() const {
+        if (ncch_data.romfs_file) {
+            return MakeResult<std::unique_ptr<FileBackend>>(std::make_unique<IVFCFile>(
+                *ncch_data.romfs_file, ncch_data.romfs_offset, ncch_data.romfs_size));
```
You should be able to omit the template parameter to `MakeResult` and let it be auto-deduced.

wwylele replied:
No, there is a `IVFCFile`->`FileBackend` pointer cast, so the type need to be there

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+            return MakeResult<std::unique_ptr<FileBackend>>(std::make_unique<IVFCFile>(
+                *ncch_data.romfs_file, ncch_data.romfs_offset, ncch_data.romfs_size));
+        } else {
+            LOG_ERROR(Service_FS, "Unable to read RomFS!");
```
Isn't this a pretty common thing for homebrew to try to do? (If I understand the issue reports we've had correctly.) Maybe we should turn down these log levels if this is expected.

yuriks commented on src/core/file_sys/archive_selfncch.h:
```diff
@@ -17,22 +18,29 @@
+    boost::optional<std::shared_ptr<std::vector<u8>>> icon;
+    boost::optional<std::shared_ptr<std::vector<u8>>> logo;
+    boost::optional<std::shared_ptr<std::vector<u8>>> banner;
+    boost::optional<std::shared_ptr<FileUtil::IOFile>> romfs_file;
```
I don't think the `optional`s are adding much value here to be honest. shared_ptr is already nullable so this is just introducing an extra level of optionalness. If you remove them I think the rest of the code actually stays exactly the same.

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+    std::vector<u8> buffer;
+
+    if (Loader::ResultStatus::Success == app_loader.ReadIcon(buffer))
+        ncch_data.icon = std::make_shared<std::vector<u8>>(std::move(buffer));
```
While I realize that this works properly, I had to go look it up in the std::vector reference and in the implementation of ReadIcon/Logo/Banner to verify that re-using the moved-from vector was ok. I think a `buffer.clear()` in-between each of these would help isolate them from the loader implementation, or the fact that previous contents are completely discarded should be made explicit in the AppLoader documentation.

wwylele replied:
added clear

yuriks commented on src/core/file_sys/archive_selfncch.h:
```diff
@@ -17,22 +18,29 @@
-    std::shared_ptr<FileUtil::IOFile> romfs_file;
-    u64 data_offset;
-    u64 data_size;
+    NCCHData ncch_data;
```
Is there any particular reason why this was split up to a separate struct?

wwylele replied:
So that `Archive_SelfNCCH` and `SelfNCCHArchive` can share the same struct and copying data can be simpler.

yuriks commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,254 @@
+    std::shared_ptr<FileUtil::IOFile> romfs_file_;
+    if (Loader::ResultStatus::Success ==
+        app_loader.ReadRomFS(romfs_file_, ncch_data.romfs_offset, ncch_data.romfs_size)) {
+        ncch_data.romfs_file = std::move(romfs_file_);
```
Add a blank line between the condition and the statement.

wwylele commented on src/core/file_sys/archive_selfncch.cpp:
```diff
@@ -0,0 +1,239 @@
+            std::string filename(file_path.exefs_filename.begin(), file_path.exefs_filename.end());
+            const size_t pos = filename.find('\0');
+            if (pos != std::string::npos)
+                filename.erase(pos);
```
This is the same idiom as [here](https://github.com/citra-emu/citra/blob/master/src/core/hle/service/cfg/cfg.cpp#L607) which was originally suggested by @lioncash 

Yeah I will add comments to it

yuriks replied:
I mean you can just do this:

```
auto end = std::find(file_path.exefs_filename.begin(), file_path.exefs_filename.end(), '\0');
std::string filename(file_path.exefs_filename.begin(), end);
```

I think this is better, anyway. (Duplicating the `file_path_exefs_filename` 3 times is a bit annoying because of how long it is though.) Up to you, I guess.

wwylele commented:
Addressed comments above. Update:

Lowered the log level of missing RomFS from ERROR to INFO (because it is common in homebrew application), and the level of missing ExeFS section from ERROR to WARNING (less common but we still handle it correctly)
